### ISSUE TITLE
Experiments on bmpc-jax

### ISSUE THREAD

**Issue by wertyuilife2** [ShaneFlandermeyer/bmpc-jax#4]
Recently I ran some experiments on bmpc-jax:

1. Ablation studies on three components: policy update interval, corrected entropy, and bmpc discount.
2. Found that bmpc-jax was unstable on cartpole-swingup-sparse, after aligning it with bmpc-torch, the issue was resolved.
3. Proposed an aligned version of bmpc-jax.
4. Some analysis about bmpc discount.

I'll share some results and thoughts here—it might be a bit long :)

### Policy Update Interval
I noticed that you added this parameter, so I tested policy update interval = 1 (blue line) vs. 10 (black line), 3 seeds:

<img src="https://github.com/user-attachments/assets/3adbcccd-3e6c-4279-aaf1-593f110e833e" width="800">

It looks like the black line performs worse on humanoid-run, keeping policy update interval = 1 might be the better choice.

## Corrected Entropy
I tested corrected entropy (blue line) vs. scaled entropy (black line) in bmpc-torch (3 seeds):

<img src="https://github.com/user-attachments/assets/1e732d05-7dfb-4bca-9f26-ab4d98455976" width="800">

On dog-run, scaled entropy seems to help a bit in the early training stages, but overall, the impact isn’t significant. In fact, their policy std are similar (dog-run):

<img src="https://github.com/user-attachments/assets/3caed372-340a-4e1d-b74d-5068a9fbbd5d" width="400">

From my experience, policy std in bmpc is mainly constrained by KL loss rather than entropy loss. Unless entropy coef is as large as 1e-3 or more, it barely affects policy std. This means that policy std is essentially controlled by log_std_min/max (which might not be ideal).

## BMPC Discount
I tested different bmpc discount in bmpc-torch (3 seeds, buffer=1M, 0.99-yellow, 0.995-green, 0.999-purple, 1.0-black):

<img src="https://github.com/user-attachments/assets/845ec489-f148-423e-a410-727e6d426817" width="800">

Similarly, I tested in bmpc-jax (3 seeds, buffer=max_steps, policy update interval=1, 0.99-black, 0.995-blue, 0.999-pink, 1.0-yellow):

<img src="https://github.com/user-attachments/assets/36d49220-7dc9-46dd-9a22-afb8cd5cf60c" width="500">

Overall, the effect of bmpc discount isn’t very noticeable, except for a improvement with discount = 0.999 on humanoid-run.

I think the imitation issue caused by reanalyze might not be significant. According to Figure 8 in the [BMPC paper](https://openreview.net/pdf?id=i7jAYFYDcM), the network policy performs almost identically to the mpc policy (with a minor gap in humanoid-run). This suggests that imitation learning is already "near perfect", leaving little room for improvement by adjusting bmpc discount.

I also compared buffer=1M (black line) vs. 2M (blue line) in bmpc-torch (3 seeds). The difference is minimal:

<img src="https://github.com/user-attachments/assets/6dfe82d2-b5f8-4e95-b2cc-558722993499" width="500">

Even with buffer=2M, the network policy can still match the mpc policy performance:

<img src="https://github.com/user-attachments/assets/7d0e46c8-39f6-4b78-9b38-573c7f29640c" width="500">

## 2. Inconsistency of bmpc-jax on cartpole-swingup-sparse
I noticed that bmpc-jax sometimes struggles to get the first reward in cartpole-swingup-sparse, whereas bmpc-torch does not. Here’s the result from bmpc-jax (5 seeds):

<img src="https://github.com/user-attachments/assets/243fe412-a609-4769-a314-d4ce6aa406e5" width="400">

Through experiments, I confirmed that this issue is unrelated to bmpc discount, policy update interval, or corrected entropy. The key difference is in how mpc selects actions:

In bmpc-torch, the final action is sampled from elite actions using a Gumbel distribution, rather than directly using the mean of the final iteration. This increases exploration in mpc policy.

Here’s the key difference in code:

bmpc-jax:
```
   # Select final action
    if deterministic:
      action = mean[..., 0, :]
    else:
      key, final_noise_key = jax.random.split(key)
      action = mean[..., 0, :] + std[..., 0, :] * jax.random.normal(
          final_noise_key, shape=batch_shape + (self.model.action_dim,)
      )
    return action.clip(-1, 1), (mean, std)
```

bmpc-torch (implemented in jax):
```
    # Select final action using Gumbel sampling
    key, gumbel_key = jax.random.split(key)
    gumbels = jax.random.gumbel(gumbel_key, shape=elite_values.shape)
    gumbel_scores = jnp.log(score) + gumbels
    action_ind = jnp.argmax(gumbel_scores, axis=-1)
    action = jnp.take_along_axis(elite_actions, action_ind[..., None, None, None], axis=-3)[..., 0, :, :]
    
    if deterministic:
      final_action = action[..., 0, :]
    else:
      key, final_noise_key = jax.random.split(key)
      final_action = action[..., 0, :] + std[..., 0, :] * jax.random.normal(
          final_noise_key, shape=batch_shape + (self.model.action_dim,)
      )    
    return final_action.clip(-1, 1), (mean, std, action)
```
After switching to Gumbel sampling, the issue was significantly reduced.

## 3. bmpc-jax-aligned
Apart from the above, there are other differences between bmpc-jax and bmpc-torch:

1. bmpc-jax's network policy imitates the mean of the final iteration, whereas bmpc-torch imitates the Gumbel-sampled mean.
2. bmpc-jax applies symlog transformation to values during planning, which is not done in tdmpc-jax or bmpc-torch.
3. bmpc-jax uses softmax with temperature=1 for action scoring, while bmpc-torch uses temperature=0.5.

To improve reproducibility, I aligned bmpc-jax with bmpc-torch by making the following changes:

1. policy update interval = 1
2. bmpc discount = 1
3. Removed symlog transformation for value (not ablated but observed bigger policy std with symlog)
4. Set softmax temperature = 0.5 for action scoring (not ablated)
5. Used Gumbel sampling for mpc action selection (also tried categorical sampling; Gumbel worked better)
6. Imitated Gumbel-sampled mean instead of the final iteration mean (worked a bit better)

Remaining differences (as far as I know):
1. Still uses corrected entropy.
2. During reanalyze, it does not modify log_std_min = -2 (tested—no significant impact).

I created a fork of bmpc-jax with these modifications, named [bmpc-jax-aligned](https://github.com/wertyuilife2/bmpc-jax-aligned). You can check my commits for the detailed changes.

Here’s a comparison between tdmpc2-jax (develop branch) and bmpc-jax-aligned across five tasks (cartpole: 5 seeds, others: 3 seeds):

<img src="https://github.com/user-attachments/assets/0cdc5b99-fb60-4580-8c9c-35ebe046b45c" width="700">

Basically reproduced the performance, but both bmpc-jax-aligned and tdmpc2-jax perform slightly worse on humanoid-run compared to the BMPC paper. Not sure why.

## 4. BMPC Discount Analysis
I noticed that the current bmpc discount implementation results in an extremely small policy loss:

- With discount = 0.99 and 0.995, policy loss is around 1e-3.
- With discount = 0.999, it’s about 1e-2.

Here’s the pi loss and pi grad norm on dog-run for discount = 0.99 (yellow), 0.999 (purple), and 1.0 (black):

<img src="https://github.com/user-attachments/assets/72bb6161-db0f-4bed-9d45-6d3f664ff963" width="400">

A very small policy loss makes entropy loss dominate, leading to noticeable differences in policy std:

<img src="https://github.com/user-attachments/assets/77156400-6e9a-4dc4-b511-3837b7a4f715" width="300">

I suspect that bmpc discount improves late-stage sample efficiency on humanoid-run mainly by increasing policy std. Meanwhile, discount = 0.99/0.995 shows no improvement likely due to its low policy loss.

To verify, I adjusted log_std_min/max in bmpc-jax-aligned:

- log_std_min: -3 -> -2, log_std_max: 1 -> -1

Here’s the result before (purple) vs. after (orange), compared to tdmpc2-jax (yellow):

<img src="https://github.com/user-attachments/assets/89644334-8c0b-49dc-8b56-333efb5eeb26" width="350">

Also tested in bmpc-torch before (black) vs. after (orange):

<img src="https://github.com/user-attachments/assets/5fc7d7ec-d723-49c1-b10c-bbcbce7656d4" width="350">

It seems like late-stage sample-efficiency drop in humanoid-run is due to low policy std, so I now recommend using:

- log_std_max: -1, log_std_min: -2

## Summary
1. policy update interval = 1 is better.
2. corrected entropy has some impact but isn’t critical.
3. bmpc discount doesn’t seem to significantly affect performance since imitation learning is already near-optimal.
4. bmpc-jax-aligned + adjusted log_std_min/max gives the best results so far -> [code](https://github.com/wertyuilife2/bmpc-jax-aligned).
5. BMPC’s policy std matters, but entropy loss doesn’t regulate it well, it’s mainly constrained by log_std_min/max (which might not be ideal).

This was a lot of data, it’s making my head spin—Let me know if I got anything wrong.

EDIT: uploaded tensorboard data [here](https://drive.google.com/file/d/1qK0GmUUyNUPp4AffwXxInmxxMB38cQUB/view?usp=sharing).

**Comment by ShaneFlandermeyer** [comment]
Nice analysis! I haven't looked through all the data in detail yet, but I can address some individual points:

* I would very much like to find a way around controlling the policy std with its log_std bounds. IMO one of the coolest things about BMPC is the ability to use only the policy without planning after training. Bounding the std to such high values potentially makes planning required for tasks that require very precise control. I've played around with using normal log std bounds in sample_actions (-5 to 1) and just clipping the std during planning, but so far have not had a ton of success. Figuring this out would be huge! 
* Regarding the value symlog during planning: I hypothesize that directly using value estimates hurts population diversity in the plan function. For example, two samples with V = 995 and 1000 produce softmax scores 0.006 and 0.993, even though they are very similar values. The symlog'd samples produce scores 0.499 and 0.501, which matches more closely with intuition. Given this it is very interesting that removing the symlog **increases** the policy std. 
* It's also interesting that the discounting doesn't make too much of a difference, as I notice a decent difference using the current code in the BMPC feature branch of tdmpc2-jax. I'll try to see if there's a discrepancy I'm missing.
* The relatively poor performance of policy update interval >1 makes sense. I mostly included that for generality and to speed up my debugging.

**Comment by wertyuilife2** [comment]
> Removed symlog transformation for value (not ablated but observed increased policy std)

Oh my bad, that's a typo. I mean I observed increased policy std with symlog.
